File: mobile/lib/models/onboarding.dart
/// Onboarding-flow models — mirror the backend Pydantic schemas at
/// backend/app/schemas/onboarding.py.
library;

enum ChatAuthorRemote { concierge, user }

class OnboardingMessage {
  const OnboardingMessage({
    required this.author,
    required this.content,
    required this.step,
    this.chips = const [],
  });

  final ChatAuthorRemote author;
  final String content;
  final String step;
  final List<String> chips;

  factory OnboardingMessage.fromJson(Map<String, dynamic> j) {
    return OnboardingMessage(
      author: (j['author'] as String) == 'user'
          ? ChatAuthorRemote.user
          : ChatAuthorRemote.concierge,
      content: j['content'] as String,
      step: j['step'] as String,
      chips: (j['chips'] as List?)?.cast<String>() ?? const [],
    );
  }
}

class StartOnboardingResponse {
  const StartOnboardingResponse({
    required this.sessionId,
    required this.welcomeMessage,
    required this.firstQuestion,
  });

  final String sessionId;
  final OnboardingMessage welcomeMessage;
  final OnboardingMessage firstQuestion;

  factory StartOnboardingResponse.fromJson(Map<String, dynamic> j) {
    return StartOnboardingResponse(
      sessionId: j['session_id'] as String,
      welcomeMessage:
          OnboardingMessage.fromJson(j['welcome_message'] as Map<String, dynamic>),
      firstQuestion:
          OnboardingMessage.fromJson(j['first_question'] as Map<String, dynamic>),
    );
  }
}

class AnswerResponse {
  const AnswerResponse({
    required this.nextStep,
    required this.conciergeReply,
    this.readbackSummary,
  });

  final String nextStep;
  final OnboardingMessage conciergeReply;
  final Map<String, dynamic>? readbackSummary;

  factory AnswerResponse.fromJson(Map<String, dynamic> j) {
    return AnswerResponse(
      nextStep: j['next_step'] as String,
      conciergeReply:
          OnboardingMessage.fromJson(j['concierge_reply'] as Map<String, dynamic>),
      readbackSummary: j['readback_summary'] as Map<String, dynamic>?,
    );
  }
}

class ReadbackConfirmResponse {
  const ReadbackConfirmResponse({
    required this.completed,
    required this.finalStep,
    required this.mandatePreview,
    required this.followUpMessage,
  });

  final bool completed;
  final String finalStep;
  final Map<String, dynamic> mandatePreview;
  final OnboardingMessage followUpMessage;

  factory ReadbackConfirmResponse.fromJson(Map<String, dynamic> j) {
    return ReadbackConfirmResponse(
      completed: j['completed'] as bool,
      finalStep: j['final_step'] as String,
      mandatePreview: j['mandate_preview'] as Map<String, dynamic>? ?? const {},
      followUpMessage:
          OnboardingMessage.fromJson(j['follow_up_message'] as Map<String, dynamic>),
    );
  }
}

File: backend/app/schemas/onboarding.py
"""Onboarding conversation — session state, questions, answers, readback.

The session is anonymous-first: no user_id until claim. Until then the session is
keyed by a server-generated session_id and lives in the in-memory store (Week 2)
or Postgres + Redis (Week 3+).

See docs/initial_specs/03_onboarding/mandate_conversation.md for the conversation script.
"""

from datetime import datetime
from enum import Enum
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from app.core.time import now_utc


class ConversationStep(str, Enum):
    """The 8 ordered steps of the Express path."""

    WELCOME = "welcome"
    Q1_GOAL = "q1_goal"
    Q2_HORIZON = "q2_horizon"
    Q3_SCENARIO_DRAWDOWN = "q3_scenario_drawdown"
    Q4_SCENARIO_REGRET = "q4_scenario_regret"
    Q5_SCENARIO_CONCENTRATION = "q5_scenario_concentration"
    Q6_MAX_DRAWDOWN = "q6_max_drawdown"
    Q7_CONSTRAINTS = "q7_constraints"
    Q8_BRIEFING = "q8_briefing"
    READBACK = "readback"
    COMPLETE = "complete"


class Author(str, Enum):
    CONCIERGE = "concierge"
    USER = "user"


class Message(BaseModel):
    """One turn in the onboarding conversation."""

    author: Author
    content: str
    step: ConversationStep
    timestamp: datetime = Field(default_factory=now_utc)
    chips: list[str] = Field(default_factory=list)


class OnboardingSession(BaseModel):
    """Anonymous onboarding session. Persists 24h or until claimed."""

    model_config = ConfigDict(use_enum_values=True)

    id: UUID = Field(default_factory=uuid4)
    started_at: datetime = Field(default_factory=now_utc)
    updated_at: datetime = Field(default_factory=now_utc)

    locale: str = "en"
    timezone: str = "UTC"
    detected_display_name: str | None = None

    current_step: ConversationStep = ConversationStep.WELCOME
    messages: list[Message] = Field(default_factory=list)
    answers: dict[str, Any] = Field(default_factory=dict)

    # Computed as we go
    risk_components_partial: dict[str, int] = Field(default_factory=dict)

    completed: bool = False
    claimed_user_id: UUID | None = None


# ── API request/response models ─────────────────────────────────────────────


class StartOnboardingRequest(BaseModel):
    locale: str = "en"
    timezone: str = "UTC"


class StartOnboardingResponse(BaseModel):
    session_id: UUID
    welcome_message: Message
    first_question: Message


class AnswerRequest(BaseModel):
    session_id: UUID
    step: ConversationStep
    answer: str  # raw user text — even chip selections submitted as text


class AnswerResponse(BaseModel):
    """Server's reply to a user's answer. Contains the next question OR the readback."""

    next_step: ConversationStep
    concierge_reply: Message
    # If readback: includes a structured summary the user can confirm/edit
    readback_summary: dict[str, Any] | None = None


class ReadbackConfirmRequest(BaseModel):
    session_id: UUID
    confirm: Literal["confirm", "edit"]
    edits: dict[str, Any] = Field(default_factory=dict)


class ReadbackConfirmResponse(BaseModel):
    completed: bool
    final_step: ConversationStep
    mandate_preview: dict[str, Any]
    follow_up_message: Message
